experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,957 @@
+          ch.valid = std::make_shared<Signal>();
+        } else {
+          // data
+          ch.valid = std::make_shared<Data>();
+        }
+        cl.valid = ch.valid;
+      }
+      if (st.ready.first) {
+        if (!st.ready.second) {
+          // signal
+          ch.ready = std::make_shared<Signal>();
+        } else {
+          // data
+          ch.ready = std::make_shared<Data>();
+        }
+        cl.ready = ch.ready;
+      }
+      if (st.data.first) {
+        if (!st.data.second) {
+          // signal
+          ch.data = std::make_shared<Signal>();
+        } else {
+          // data
+          ch.data = std::make_shared<Data>();
+        }
+        cl.data = ch.data;
+      }
+
+      circuitState[val] = cl;
+    } else {
+      if (st.valid.first) {
+        ch.valid = circuitState[val].valid.lock();
+      }
+      if (st.ready.first) {
+        ch.ready = circuitState[val].ready.lock();
+      }
+      if (st.data.first) {
+        ch.data = circuitState[val].data.lock();
+      }
+    }
+  }
+};
+
+int main(int argc, char **argv) {
+
+  InitLLVM y(argc, argv);
+
+  cl::ParseCommandLineOptions(
+      argc, argv,
+      "MLIR Standard dialect runner\n\n"
+      "This application executes a function in the given MLIR module\n"
+      "Arguments to the function are passed on the command line and\n"
+      "results are returned on stdout.\n"
+      "Memref types are specified as a comma-separated list of values.\n");
+
+  auto fileOrErr = MemoryBuffer::getFileOrSTDIN(inputFileName.c_str());
+  if (std::error_code error = fileOrErr.getError()) {
+    errs() << argv[0] << ": could not open input file '" << inputFileName
+           << "': " << error.message() << "\n";
+    return 1;
+  }
+
+  // Load the MLIR module.
+  mlir::MLIRContext context;
+  context.loadDialect<func::FuncDialect, memref::MemRefDialect,
+                      handshake::HandshakeDialect>();
+
+  // functions feeding into HLS tools might have attributes from high(er)
+  // level dialects or parsers. Allow unregistered dialects to not fail in
+  // these cases.
+  context.allowUnregisteredDialects();
+
+  // functions feeding into HLS tools might have attributes from high(er)
+  // level dialects or parsers. Allow unregistered dialects to not fail in
+  // these cases.
+  context.allowUnregisteredDialects();
+
+  SourceMgr sourceMgr;
+  sourceMgr.AddNewSourceBuffer(std::move(*fileOrErr), SMLoc());
+  mlir::OwningOpRef<mlir::ModuleOp> module(
+      mlir::parseSourceFile<ModuleOp>(sourceMgr, &context));
+  if (!module)
+    return 1;
+
+  ////////////////////////////////////////////////////////////////////////
+
+  auto *modOp = module->getOperation();
+  auto &region = modOp->getRegions().front();
+  auto &block = region.getBlocks().front();
+  auto funcOp = *block.getOps<handshake::FuncOp>().begin();
+
+  Circuit circ(funcOp);
+
+  // iterating througth all the ops of the kernel function
+  for (Operation &op : funcOp.getOps()) {
```

Comment:
I really don't like this, and I don't understand why we're going for something so different from the level of simplicity I outlined in the spec and which I am 99% sure is achievable. 

```cpp
if (auto muxOp = dyn_cast<handshake::MuxOp>(op)) {
    // Constructs an instance of MuxModel which will be associated
    // to muxOp during simulation of funcOp
    simulator.registerModel<MuxModel>(muxOp);
}
```

Let's take `MulIOp` as an example, I don't understand why any of the manual lhs/rhs/result initialization work is necessary. The `Circuit` class should be able to do most of this automatically with the *same piece of generic code for all possible components*. The only two things that need to be per-component, and that should be part of the execution models themselves (potentially in their constructor), are the following.

1. The initialization of the component's internal state (as if `reset` was active), and
2. the initialization of the component's outputs (as if `reset` was active).